kotlin/workflow-ui-android/src/main/java/com/squareup/workflow/ui/backstack/BackStackContainer.kt

```diff
@@ -51,45 +60,45 @@ class BackStackContainer(
 
   private val showing: View? get() = if (childCount > 0) getChildAt(0) else null
 
-  fun takeScreens(
-    screens: Observable<out BackStackScreen<*>>,
-    viewRegistry: ViewRegistry
-  ) {
-    takeWhileAttached(screens.distinctUntilChanged { a, b -> a.key == b.key }) {
-      show(it, screens, viewRegistry)
-    }
-  }
+  private lateinit var registry: ViewRegistry
 
-  private fun show(
-    newScreen: BackStackScreen<*>,
-    screens: Observable<out BackStackScreen<*>>,
-    viewRegistry: ViewRegistry
-  ) {
-    showing?.let { if (it.backStackKey == newScreen.key) return }
+  private fun update(newValue: BackStackScreen<*>) {
+    // Existing view is of the right type, just update it.
+    showing
+        ?.takeIf { it.hasRunnerFor(newValue.wrapped) }
+        ?.updateRunner(newValue.wrapped)
+        ?.also { return }
 
-    val updateTools = viewStateStack.prepareToUpdate(newScreen.key)
+    val updateTools = viewStateStack.prepareToUpdate(newValue.key)
+    val newView = registry.buildView(newValue.wrapped, this)
+        .apply { updateTools.setUpNewView(this) }
 
+    // Showing something already, transition with push or pop effect.
     showing
-        ?.let {
-          updateTools.saveOldView(it)
-          viewRegistry.getEffect(it.backStackKey, newScreen.key, updateTools.direction)
-              .execute(
-                  from = it,
-                  to = newScreen,
-                  screens = screens,
-                  viewRegistry = viewRegistry,
-                  container = this,
-                  setUpNewView = updateTools::setUpNewView,
-                  direction = updateTools.direction
-              )
+        ?.let { oldView ->
+          updateTools.saveOldView(oldView)
+
+          val newScene = Scene(this, newView)
+
+          val (outEdge, inEdge) = when (PUSH) {
+            updateTools.direction -> Pair(Gravity.START, Gravity.END)
```

Comment:
This looks unidiomatic to me, usually the variable is in the when and the cases are constant - is something going on here I'm not seeing?

Reply:
Whoa,  that ain't right.